lib/graphql/schema.rb

```diff
@@ -1112,7 +1112,13 @@ def possible_types(type = nil, context = GraphQL::Query::NullContext)
           if type.kind.union?
             type.possible_types(context: context)
           else
-            own_possible_types[type.graphql_name] ||
+            visible_possible_types = own_possible_types[type.graphql_name]&.select do |possible_type|
+              next true unless type.kind.interface?
+              next true unless possible_type.kind.object?
+
+              possible_type.interfaces(context).any? { |interface| interface.graphql_name == type.graphql_name }
```

Comment:
I had to use `graphql_name` for comparison here rather than just the type because of: https://github.com/rmosolgo/graphql-ruby/blob/master/spec/graphql/interface_type_spec.rb#L10

This broke old tests because the own possible types would return `GraphQL::Schema::Interface` objects, while we were passing in a `GraphQL::InterfaceType` class to the above `possible_types` method (this also failed for a few other tests in this file). 

If we don't care about these tests as this is legacy behaviour then I would suggest changing this line to:

`possible_type.interfaces(context).include?(type)`

Reply:
Yeah, I doubt it's much overhead to keep legacy support in there. Let's just add a comment so that we _might_ remember to simplify it once it's possible: 

```suggestion
              # Use `.graphql_name` comparison to match legacy vs class-based types. 
              # When we don't need to support legacy `.define` types, use `.include?(type)` instead.
              possible_type.interfaces(context).any? { |interface| interface.graphql_name == type.graphql_name }
```